### PR TITLE
API client meta-ish

### PR DESCRIPTION
**Pull request by brianswko**

### CODE REVIEW

**Comment by brianswko** [comment]
@cfurrow what do you think of this approach vs. https://github.com/lumoslabs/automatic-client/pull/1?

File: lib/automatic/api_client/client_delegator.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  class ApiClient
+    class ClientDelegator
+      attr_reader :api_client, :delegated_class
+
+      def initialize(api_client, delegated_class)
+        @api_client = api_client
+        @delegated_class = delegated_class
+      end
+
+      def method_missing(method_sym, *arguments, &block)
```
Comment (cfurrow) [inline lib/automatic/api_client/client_delegator.rb]: If you're going to do a `method_missing` you should also add a `respond_to?` handler. Then you can change line 12 to be `super unless respond_to?(method_sym)`

File: lib/automatic/api_client/client_delegator.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+        super unless delegated_class.respond_to?(method_sym)
+
+        arguments[1 + delegated_class.method(method_sym).arity] ||= {}
```
Comment (cfurrow) [inline lib/automatic/api_client/client_delegator.rb]: This looks pretty hairy. Hard to follow what it's doing.
- take the incoming arguments (assuming that they are non-nil)
- add 1 to the `delegated_class` method's expected argument size to get the argument index
- if that argument index is nil, set it to `{}`, otherwise, leave it alone.
- then, take that argument hash, and set the `client` key.

It assumes a lot, and feels brittle. Have you thought about alternate ways of doing this? `method_missing` may be too "magical" for this purpose.

Comment (cfurrow) [inline lib/automatic/api_client/client_delegator.rb]: I do like how terse the `trips.rb` and `vehicles.rb` are, however.

File: lib/automatic/api_client.rb [hunk 3]
```diff
@@ -0,0 +1,95 @@
+require 'automatic/api_client/trips'
+require 'automatic/api_client/vehicles'
+
+module Automatic
+  class ApiClient
```
Comment (cfurrow) [inline lib/automatic/api_client.rb:5]: Seems to be a large amount of duplication here, with regards to the existing `Automatic::Client`. Have you thought about ways of sharing behavior between the two, and modifying only what is required to be different?

Seems like the main branching point is the instance-level auth token. `Automatic::Client` sets it at the class-level, and `Automatic::ApiClient` sets it at the instance-level. Otherwise, the methods seem to perform the same actions, regardless of being class or instance methods. DRYing that up would be awesome. Did you have thoughts on how you would go about that? Perhaps some kind of service model that could facilitate this?

Comment (brianswko) [inline lib/automatic/api_client.rb:5]: yeah since these same methods are class-level in one and instance in another...does creating a module which is extended in `Automatic::Client` and included in `Automatic::ApiClient` make sense?

File: lib/automatic/api_client/client_delegator.rb (hunk 2 above)
Comment (brianswko) [inline lib/automatic/api_client/client_delegator.rb]: I wasn't sure how else to insert a key-value into the options hash when calling the method in the delegated_class. `arguments.last` doesn't work since the options hash is optional, and the methods also have varying numbers of arguments.

The alternative is to make `trips.rb` and `vehicles.rb` more verbose and know which individual methods in `Automatic::Models::Trips` need to have the instance-level client passed in. Do-able, but that ties these classes together closer than I'd like. Very open to other potential solutions if you've got any

File: lib/automatic/api_client.rb (hunk 3 above)
Comment (cfurrow) [inline lib/automatic/api_client.rb:5]: That may be an alright solution.

File: lib/automatic/api_client/client_delegator.rb (hunk 2 above)
Comment (cfurrow) [inline lib/automatic/api_client/client_delegator.rb]: My initial thought was to change this gem to move to an instance-based approach, and gut the entire class-level stuff. You could also create some kind of singleton class that handles the existing case of "only one client, and it's at the class level" to handle the legacy stuff. If that makes sense.

Comment (brianswko) [inline lib/automatic/api_client/client_delegator.rb]: Yeah a lot of this was driven by the desire to maintain as much backwards compatibility as possible. Will see what I can do.

**Comment by brianswko** [comment]
@cfurrow did some more refactoring of this, :eyes:?

**Comment by cfurrow** [comment]
I like that more. Thanks!

llama
